### DISCUSSION TITLE
Trying to understand how semaphore uses repositories

### DISCUSSION THREAD

**Discussion by lockiime** [semaphoreui/semaphore#1552] [Q&A]
Hello.. noob question.

Does Semaphore have to clone the repository each time a new Task Template is created?

I have only one repository with a bunch of playbooks and roles.  It doesn't make sense to me that each task created clones the same repository when it should just be refreshing it?  Maybe I didn't set it up correctly.  Can someone help explain?

Thanks in advance.

**Comment by hmoffatt** [comment]
I can think of two reasons that each has a unique check out per task template:

- each one can use a different repository
- you can run multiple tasks at once, so they each need their own checkout or else what if the repository has changed since the first task started?

Anyway it just uses some extra disk space, does it matter?

**Reply by lockiime** [reply]
Thanks for your answer hmoffatt.  I have looked it over and came to the conclusion that is just how it works.  It seems wasteful to checkout the same repository to create a task that uses a different playbook in the same repository.  I have files and packages that are stored in the repository also.  currently only 2.6GB times however many templates I will create will add up unnecessarily.  Let me know if there is a way to point to the same checked out repo instead of duplicating it for each task template.  I'd rather it just refresh the latest changes.  I would want it to always run the latest anyway when it runs.

Thank you for your answer.  I might try awx and see how it is handled.

**Reply by hmoffatt** [reply]
Yes I see it's a nuisance with 2.6Gb of files, I only have a few megabytes.

I don't think it can be safe for Semaphore to share the checkout in general though. If you run a task, and while it's running you update the repository and start a different task, it has to make a new checkout somewhere.
